chunky/src/java/se/llbit/util/PositionalInputStream.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2021 Chunky Contributors
+ *
+ * This file is part of Chunky.
+ *
+ * Chunky is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Chunky is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * You should have received a copy of the GNU General Public License
+ * along with Chunky.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package se.llbit.util;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.function.LongConsumer;
+
+public class PositionalInputStream extends InputStream {
+    private InputStream stream;
+    private LongConsumer update;
+    private long count = 0;
+    private long mark;
+    private long markLimit;
+    private boolean canMark;
+
+    public PositionalInputStream(InputStream stream, LongConsumer update) {
+        this.stream = stream;
+        this.update = update;
+        this.canMark = stream.markSupported();
+    }
+
+    public long getPosition() {
+        return count;
+    }
+
+    @Override
+    public int available() throws IOException {
+        return stream.available();
+    }
+
+    @Override
+    public void close() throws IOException {
+        stream.close();
+    }
+
+    @Override
+    public void mark(int readlimit) {
+        if (canMark) {
+            stream.mark(readlimit);
+            mark = count;
+            markLimit = mark + readlimit;
+        }
+    }
+
+    @Override
+    public boolean markSupported() {
+        return canMark;
+    }
+
+    @Override
+    public int read() throws IOException {
+        update.accept(++count);
+        return stream.read();
+    }
+
+    @Override
+    public int read(byte[] b) throws IOException {
+        int actual = stream.read(b);
+        count += actual;
+        update.accept(count);
+        return actual;
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        int actual = stream.read(b, off, len);
+        count += actual;
+        update.accept(count);
+        return actual;
+    }
+
+    @Override
+    public void reset() throws IOException {
+        if (canMark && count <= markLimit) {
+            stream.reset();
+            count = mark;
+        }
```

Comment:
This doesn't reset `count` if `stream.markSupported()` returns `false`.

Reply:
Count should not be reset if this is called and mark is not supported. ([Actually it should probably throw an IOException](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#reset--))

Reply:
I changed `reset` to throw an `IOException` if `mark` has not been called, more bytes have been read than `readlimit`, or if `markSupported` returns `false`.

In these cases, the position of the stream (and if an `IOException` is thrown) depends on the underlying stream so the position is unpredictable. (This is the strictest implementation of the [`InputStream.reset` contract](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#reset--))

Reply:
Thank you :+1: This should ensure that we don't get any surprises when using this stream.